applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/shell_thin_adjoint_element_3D3N.cpp

```diff
@@ -0,0 +1,850 @@
+
+    for(SizeType i = 0; i < geom.size(); i++)
+    {
+        int index = i * 6;
+        NodeType & iNode = geom[i];
+
+        rResult[index]     = iNode.GetDof(ADJOINT_DISPLACEMENT_X).EquationId();
+        rResult[index + 1] = iNode.GetDof(ADJOINT_DISPLACEMENT_Y).EquationId();
+        rResult[index + 2] = iNode.GetDof(ADJOINT_DISPLACEMENT_Z).EquationId();
+
+        rResult[index + 3] = iNode.GetDof(ADJOINT_ROTATION_X).EquationId();
+        rResult[index + 4] = iNode.GetDof(ADJOINT_ROTATION_Y).EquationId();
+        rResult[index + 5] = iNode.GetDof(ADJOINT_ROTATION_Z).EquationId();
+    }
+}
+
+void ShellThinAdjointElement3D3N::GetDofList(DofsVectorType& ElementalDofList, ProcessInfo& CurrentProcessInfo)
+{
+    const SizeType num_dofs = GetNumberOfDofs();
+    ElementalDofList.resize(0);
+    ElementalDofList.reserve(num_dofs);
+
+    GeometryType & geom = this->GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); i++)
+    {
+        NodeType & iNode = geom[i];
+
+        ElementalDofList.push_back(iNode.pGetDof(ADJOINT_DISPLACEMENT_X));
+        ElementalDofList.push_back(iNode.pGetDof(ADJOINT_DISPLACEMENT_Y));
+        ElementalDofList.push_back(iNode.pGetDof(ADJOINT_DISPLACEMENT_Z));
+
+        ElementalDofList.push_back(iNode.pGetDof(ADJOINT_ROTATION_X));
+        ElementalDofList.push_back(iNode.pGetDof(ADJOINT_ROTATION_Y));
+        ElementalDofList.push_back(iNode.pGetDof(ADJOINT_ROTATION_Z));
+    }
+}
+
+int ShellThinAdjointElement3D3N::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    GeometryType& r_geom = GetGeometry();
+
+    // verify that the variables are correctly initialized
+    KRATOS_CHECK_VARIABLE_KEY(DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(ROTATION);
+    KRATOS_CHECK_VARIABLE_KEY(VELOCITY);
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+    KRATOS_CHECK_VARIABLE_KEY(DENSITY);
+    KRATOS_CHECK_VARIABLE_KEY(SHELL_CROSS_SECTION);
+    KRATOS_CHECK_VARIABLE_KEY(THICKNESS);
+    KRATOS_CHECK_VARIABLE_KEY(CONSTITUTIVE_LAW);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_DISPLACEMENT);
+    KRATOS_CHECK_VARIABLE_KEY(ADJOINT_ROTATION);
+
+    // check properties
+    KRATOS_ERROR_IF(this->pGetProperties() == NULL) << "Properties not provided for element " << this->Id() << std::endl;
+
+    const PropertiesType & props = this->GetProperties();    
+
+    if(props.Has(SHELL_CROSS_SECTION)) // if the user specified a cross section ...
+    {
+        const ShellCrossSection::Pointer & section = props[SHELL_CROSS_SECTION];
+        KRATOS_ERROR_IF(section == NULL) << "SHELL_CROSS_SECTION not provided for element " << this->Id() << std::endl;
+
+        section->Check(props, r_geom, rCurrentProcessInfo);
+    }
+    else // ... allow the automatic creation of a homogeneous section from a material and a thickness
+    {
+        KRATOS_ERROR_IF_NOT(props.Has(CONSTITUTIVE_LAW)) << "CONSTITUTIVE_LAW not provided for element " << this->Id() << std::endl;
```

Comment:
If we call here the checks from the base element it will also check for the primal dofs (DISPLACEMENT and ROTATION) which doesn't exist at the time when calling the check of the adjoint elements.

Reply:
I think the better solution here is to define the checks in separate functions in the base class so you can reuse selected ones here.

Reply:
I separated the check-functions of the base classes. @KlausBSautter I hope it is okay for you that I changed the check-function of your CrBeamElement3D2N?